### DISCUSSION TITLE
Recursive type

### DISCUSSION THREAD

**Discussion by w14** [PyO3/pyo3#4216] [General]
How do you pass a recursive type to Python?

For example, PyO3's #[pyclass] seems to fail:

```rust
#[pyclass]
enum T {
    Empty {},
    T { value: Vec<T> },
}
```

**Comment by davidhewitt** [comment]
I think what you probably want is to have the recursive value wrapped in `Py<T>`, i.e. a Python `T` object:

```
#[pyclass]
enum T {
    Empty {},
    T { value: Vec<Py<T>> },
}
```

**Reply by warriorstar-orion** [reply]
I know this discussion is about 6 months old but I was trying to implement a similar recursive enum type and the compiler doesn't seem happy with the suggested pattern. In my case I'm attempting something like:

```rust
#[pyclass]
#[derive(Clone)]
pub enum Constant {
    Null(),
    Number(u64),
    String(String),
}

#[pyclass]
pub enum Expression {
    Constant{ inner: Constant },
    BinaryOp{
        op: BinaryOperator,
        lhs: Py<Expression>,
        rhs: Py<Expression>,
    },
}
```

```
error[E0277]: the trait bound `pyo3::Py<foo::bar::Expression>: std::clone::Clone` is not satisfied
  --> src\foo\bar.rs:15:1
   |
15 | #[pyclass]
   | ^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `pyo3::Py<foo::bar::Expression>`
   |
   = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)
```

EDIT: I see there's a Clone trait implemented for `Py<T>` included with the `py-clone` feature, but I assume it's not part of the default feature set for a reason, reading the docs about it now.

**Reply by Icxolu** [reply]
I opened a fix for this in #4694.

**Reply by ChristopherRabotin** [reply]
For future readers, in version 0.25.1 and earlier, this requires explicitly writing out the name of the enum, cf. https://github.com/PyO3/pyo3/issues/5420.
